### PR TITLE
Client destroy

### PR DESCRIPTION
**Pull request by maksimf**
- [x] add logs when that method gets called
- [x] test responses of calling get_treatment and manager methods after the call to shutdown happened

### CODE REVIEW

**Comment by maksimf** [comment]
Logs after `split_client.destroy` called:

```
5 splits retrieved. since=-1
storing split (ruby)
storing split (ruby_5)
storing split (ruby_4)
storing split (ruby_3)
storing split (ruby_1)
segments seen(0): []


Started GET "/destroy" for ::1 at 2017-06-15 18:45:44 +0300
  ActiveRecord::SchemaMigration Load (0.3ms)  SELECT "schema_migrations".* FROM "schema_migrations"
Processing by TestingRubyController#destroy as HTML
Shutting down split client...
Split client shutting down complete
```

**Comment by maksimf** [comment]
@patricioe Ready to be merged

**Comment by patricioe** [comment]
Are impressions being sent during the destroy process?

File: lib/splitclient-rb/clients/split_client.rb [hunk 1]
```diff
@@ -116,6 +116,19 @@ def get_treatment(
       parsed_treatment(multiple, treatment_data)
     end
 
+    def destroy
+      @config.logger.info('Split client shutdown started...') if @config.debug_enabled
+
+      @config.threads.each { |name, thread| Thread.kill(thread) }
```
Comment (patricioe) [inline lib/splitclient-rb/clients/split_client.rb]: don't we need to send the impressions before shutdown?

**Comment by maksimf** [comment]
@patricioe nope, all left impressions are dropped

File: lib/splitclient-rb/clients/split_client.rb (hunk 1 above)
Comment (patricioe) [inline lib/splitclient-rb/clients/split_client.rb]: @maksimf can we fire a last call before shutdown and make it NON blocking?

Comment (maksimf) [inline lib/splitclient-rb/clients/split_client.rb]: @patricioe yes, will do

File: lib/splitclient-rb/cache/adapters/memory_adapters/queue_adapter.rb [hunk 2]
```diff
@@ -10,6 +10,11 @@ def initialize(max_size)
             @current_size = Concurrent::AtomicFixnum.new(0)
           end
 
+          def clear(_ = nil)
+            @queue = Queue.new
+            @current_size = Concurrent::AtomicFixnum.new(0)
```
Comment (patricioe) [inline lib/splitclient-rb/cache/adapters/memory_adapters/queue_adapter.rb]: why are you assigning a new variable? if the existing variable is atomic, why not setting it to zero?

File: lib/splitclient-rb/cache/senders/impressions_sender.rb [hunk 3]
```diff
@@ -31,13 +31,19 @@ def call
...
-            @config.logger.info('Starting impressions service')
+          @config.threads[:impressions_sender] = Thread.new do
+            begin
+              @config.logger.info('Starting impressions service')
 
-            loop do
+              loop do
+                post_impressions
+
+                sleep(SplitIoClient::Utilities.randomize_interval(@config.impressions_refresh_rate))
```
Comment (patricioe) [inline lib/splitclient-rb/cache/senders/impressions_sender.rb]: @maksimf remember that this cause an infinite loop before. If the post errors out, it skips the sleep and cycles again there is a nextwork issue. Please restablish the previous pattern.

File: lib/splitclient-rb/cache/senders/impressions_sender.rb [hunk 4]
```diff
@@ -31,13 +31,19 @@ def call
...
+              end
+            rescue SplitIoClient::ImpressionShutdownException
               post_impressions
 
-              sleep(SplitIoClient::Utilities.randomize_interval(@config.impressions_refresh_rate))
+              @impressions_repository.clear
```
Comment (patricioe) [inline lib/splitclient-rb/cache/senders/impressions_sender.rb]: You are accounting for the new impressions listener shutdown as well?

Comment (maksimf) [inline lib/splitclient-rb/cache/senders/impressions_sender.rb]: Nope, are you ok if I'll fix that after Impression listener PR is merged?

File: lib/splitclient-rb/cache/senders/impressions_sender.rb (hunk 3 above)
Comment (maksimf) [inline lib/splitclient-rb/cache/senders/impressions_sender.rb]: @patricioe But we're catching exceptions inside `post_impressions` method:

```ruby
def post_impressions
  impressions_client.post
rescue StandardError => error
  @config.log_found_exception(__method__.to_s, error)
end
```

**Comment by maksimf** [comment]
@patricioe Done

File: lib/splitclient-rb/cache/senders/impressions_sender.rb (hunk 4 above)
Comment (patricioe) [reply lib/splitclient-rb/cache/senders/impressions_sender.rb]: @maksimf I'm merging this. Mind addressing the impression listener shutdown right now? I'm merging this now.

**Comment by patricioe** [comment]
merging this.

@maksimf you are missing the changelog entry.

**Pull request merged**